src/Synapse.ts

```diff
@@ -0,0 +1,68 @@
+type Logger<T> = (oldState: StoreState<T>, newState: StoreState<T>) => void;
+// @ts-ignore
+type StoreState<C> = InstanceType<C[keyof C] & { state: object }>['state'];
+// @ts-ignore
```

Comment:
I can't find a way around this. I'm getting this annoying error, even though the resulting type is correct. Halp. 

```ts
Type '"state"' cannot be used to index type 'InstanceType<C[K]>'.
```